### PR TITLE
Add test framework and example test

### PR DESCRIPTION
**Pull request by IanVS**
Tasks before merge-ready:

 - [X] Switch from Grunt to npm test
 - [X] Add a test to test the test framework
 - [ ] [Stretch Goal] Extract into an npm module

Added:

- [Mocha](http://mochajs.org/)
- [Sinon](http://sinonjs.org/)
- [Chai](http://chaijs.com/)

To execute tests, `npm test` inside server directory.

So far, only one test added to show that it works.

### CODE REVIEW

File: server/tasks/register/test.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+module.exports = function (grunt) {
+	grunt.registerTask('test', [
+		'mochaTest'
+	]);
+};
```
Comment (kriswill) [inline server/tasks/register/test.js]: careful with those tabs!

Comment (IanVS) [inline server/tasks/register/test.js]: Good catch. That's what I get for copying from an example.  Thought I had sublime set up to to replace them. Thanks again.

**Comment by Globegitter** [comment]
@IanVS Another great contribution. Instead of grunt though, do you think you can just make it in such a way that tests are run via `npm test`? Similar to how sane does it? https://github.com/artificialio/sane/blob/master/package.json#L8

**Comment by IanVS** [comment]
Sure I can try doing it that way.  Reason I used grunt was partly because Sails uses it natively for its task running, partly because I wanted to get a little familiar with how to set up grunt tasks, and the biggest reason...that's how it was done in the tutorial I was following.  :)

I'll try refactoring tonight.

**Comment by Globegitter** [comment]
Yeah grunt is the old way :P Sails does support gulp now as well I believe, which is better than grunt but yeah I think it still is a bit too much overhead just for running tests :) 

But that is good stuff.

**Comment by mgenev** [comment]
Hey, @IanVS this is awesome! I will review when i've had some sleep. I'm not sure if I showed you my list of suggested features for SANE, but the test framework was towards the top, so this is pretty perfect

**Comment by IanVS** [comment]
Yeah I wanted to start playing around with adding functionality and tweaks, and wanted to make sure I didn't break any of your code.  I also plan to give BDD a shot, even though I really have only the faintest clue what it is and how to do it.  Learning as I go here.

I'll be updating the tests tonight to use straight npm as @Globegitter suggested.  Looks very straight forward to do.  So don't review just yet.

I'm also investigating git hooks to prevent tabs from being committed/pushed, as @kriswill called me out on.

**Comment by IanVS** [comment]
By the way, if I know there's additional work to be done on a PR, is it best practice to close the PR for the time being, until issues are addressed and I think it's ready to merge?

**Comment by kriswill** [comment]
@IanVS ember-cli has .jshintrc tests in their standard suite.  see https://github.com/ember-cli/ember-cli/blob/master/tests/unit/jshint-test.js

**Comment by Globegitter** [comment]
There are different conventions per project, but usually not closing the PR. You could either just mention **Not ready to merge** at the top if you know from the start it is not ready, and/or a checklist with a list of things that already has been done and things that are still missing. Then it really just has to pass at least one of the maintainers code review and tests if any are present.

**Comment by mgenev** [comment]
@IanVS for BDD check out this https://github.com/LateRoomsGroup/moonraker I use it at work and like it a lot

**Comment by IanVS** [comment]
@mgenev, moonraker looks cool, but I guess my goal is to stick fairly closely to the syntax used in ember-cli and sane (although they're much more sophisticated than this is so far).

@Globegitter I'm sorry but I have no idea how machinepacks work or how to make this an npm module.  Maybe that's something we can do later?  I'd like to try to do more work on making tests and improving auth.

**Comment by kriswill** [comment]
I think it's a good start!

I think what @Globegitter would like to see, is how we can leverage the work in this project to integrate with sane-cli's addon system (which is still very alpha).

Eventually we would like to expose rich enough interfaces that nearly *every piece* of the Sane stack is implemented using the addon system.  This would give us immense flexibility to create new features, and allow the community to contribute alternative capabilities into the stack, like new deployment targets, configuration systems, alternate testing frameworks, new code generator artifacts, etc.  The future is bright!

**Comment by mgenev** [comment]
@IanVS with that cucumber/yadda, since it ends up in plain English,  I set it up so our  QA staff that doesn't code, writes out the behaviors of the products we get and I don't have to worry about that part :D

**Comment by Globegitter** [comment]
@IanVS yes sorry I should have been a bit more specific, this where more thoughts for the future. I have now learned to write down ideas immediately somewhere :) What you have right now is a really good and I would say LGTM (let's get this merged) if @mgenev does not have any objections.

And yes @kriswill is explaining the core idea in a bit more detail. But again nothing to worry for now.

**Comment by Globegitter** [comment]
And regards BDD and moonraker haven't really done that myself yet, but I think these are things we can tackle in future PRs as well. It is good to have this basis for now.
